libkernelflinger/log.c

```diff
@@ -163,11 +170,51 @@ void vlog(const CHAR16 *fmt, va_list args)
 void log(const CHAR16 *fmt, ...)
 {
 	va_list args;
+#ifdef KERNELFLINGER_LOG_FILE
+	EFI_FILE *root;
+	const CHAR16 *filename = L"\\kernelflinger.log";
+	UINTN length;
+	EFI_STATUS ret;
+#endif
 
+#ifndef KERNELFLINGER_LOG_FILE
 	if (!serial && EFI_ERROR(serial_init()))
 		return;
 
 	va_start(args, fmt);
 	vlog(fmt, args);
 	va_end(args);
+#else
+		va_start(args, fmt);
+		length = VSPrint(buf16, sizeof(buf16), (CHAR16 *)fmt, args) + 1;
+
+		if (EFI_ERROR(str_to_stra(buf8, buf16, length)))
+			return;
+		length--;
+
+		if (buf_pos + length >= LOGFILE_BUF_SIZE ) {
+			length = (LOGFILE_BUF_SIZE ) - buf_pos;
+		}
+
+		memcpy(data_buf + buf_pos, buf8, length);
+		buf_pos += length;
+
+		if(buf_pos == 0)
+			return;
+
+		if(OUTPUT_KLOG_ZERO == 1){
```

Comment:
coding style: there must a space before '(' and after ')'.  This is a mistake you made almost everywhere !

Reply:
Already modify. There is a space before '(' and  after ')'.